### ISSUE TITLE
Improve workflow for requesting reviews

### ISSUE THREAD

**Issue by gsuess** [Reviewable/Reviewable#598]
use case: As an author of a pull request I want reviewable to re-request a review from the reviewer when I address the discussions of the code review so that the PR will show-up on the github issue filter "review-request" for that respective reviewer.

So basically fixing something that was discussed in the code-review I come back to reviewable and mark it as done. At this point reviewable should ask me if the reviewer who has originally posted the discussion should be requested to review it again. If yes, add the respective reviewer to the reviewers list on github.

**Comment by pkaminski** [comment]
Thanks for suggesting this, I hadn't heard of this particular workflow yet!  I think most people treat the requested reviewers as an initial request only, then rely on reply notifications to come back to the review when needed.

Do GitHub reviews re-request reviewers when an author responds to comments?  (AFAIK they don't, but wanted to double-check if I missed something.)

Is there a reason you're using the GitHub PR list rather than Reviewable's review dashboard?  Reviewable will automatically figure out which reviews need your attention without needing to set the requested reviewer flag, once the review has started.

**Comment by gsuess** [comment]
> Do GitHub reviews re-request reviewers when an author responds to comments? (AFAIK they don't, but wanted to double-check if I missed something.)

I would have to double check, but I think they don't either. Which would mean that reviewable could have another advantage to github.

> Is there a reason you're using the GitHub PR list rather than Reviewable's review dashboard? Reviewable will automatically figure out which reviews need your attention without needing to set the requested reviewer flag, once the review has started.

Fair point. However here are some more use-cases:

* Send reminders to reviewers via slack: i.e. https://pullreminders.com
* Measure turnaround times for reviews to be completed.

Or in more general sense: when Github is the platform and the platform holds the state on who needs to review something then Reviewable, being on the platform, should maintain that state. Especially when it already does Approve or Reject PRs, which clears the reviewers.

**Comment by pkaminski** [comment]
> Or in more general sense: when Github is the platform and the platform holds the state on who needs to review something then Reviewable, being on the platform, should maintain that state.

Indeed, but I'm questioning whether that's what requested reviewers are for.  Given the lack of a formal definition from GitHub, I'm inferring from their own site's behavior that it's intended to get a review started, and not to keep it going.  In general, lacking a compelling argument to the contrary, I prefer to match GitHub's behavior to reduce surprise.

Let's see if we can find a compelling argument:

> Send reminders to reviewers via slack: i.e. https://pullreminders.com

That is admittedly pretty compelling.  Does it only trigger on review requests to figure out who a review is waiting on?  If so, how does it deal with waiting on the PR author, who can't be added as a requested reviewer?

> Measure turnaround times for reviews to be completed.

I feel this is less compelling, both intrinsically and because you can do it by parsing the messages that Reviewable sends out, which include the list of people now being waited on.  (One customer actually does this.)

**Comment by gsuess** [comment]
> Indeed, but I'm questioning whether that's what requested reviewers are for. Given the lack of a formal definition from GitHub, I'm inferring from their own site's behavior that it's intended to get a review started, and not to keep it going. In general, lacking a compelling argument to the contrary, I prefer to match GitHub's behavior to reduce surprise.

Well I think the fact that they have a filter that they call "Awaiting review from you" which will not show up any PRs that do not have the reviewer assigned, means that at least part of GitHub expects reviewers to be re-assigned on each review cycle.

>  Does it only trigger on review requests to figure out who a review is waiting on? 

yes

> If so, how does it deal with waiting on the PR author, who can't be added as a requested reviewer?

If there is no reviewer assigned, then its waiting on the author.

>I feel this is less compelling, both intrinsically and because you can do it by parsing the messages that Reviewable sends out, which include the list of people now being waited on. (One customer actually does this.)

pullreminders does measure these metrics but knows nothing about reviewable. It accesses it from a single-source of truth from GitHub. Sure, we could build a custom solution, but that solution would be very specific to reviewable.

**Comment by abinoda** [comment]
@pkaminski Hey! I'm the creator of Pull Reminders. Pull Reminders obviously uses the GitHub API, but I've run into a good number of teams that use Pull Reminders in conjunction with Reviewable. I'm not familiar with the internals of Reviewable, but I'm guessing you also use GitHub's API for things like review requests?

>  Given the lack of a formal definition from GitHub, I'm inferring from their own site's behavior that it's intended to get a review started, and not to keep it going. In general, lacking a compelling argument to the contrary, I prefer to match GitHub's behavior to reduce surprise.

This is not GitHub's intention. I just spoke with a designer at GitHub (@califa) and they are aware that the workflow for re-requesting reviews is not discoverable or intuitive. I don't want to speak on behalf of GitHub, but they mentioned maybe prioritizing adding a "Re-request review" button in the pull request UI to solve the current state of ambiguity and help make it easy for teams to employ a consistent workflow. **It'd be great if Reviewable could do this too!**

Here's a write-up of this issue which I've provided to GitHub:
https://paper.dropbox.com/doc/Re-requesting-reviews-on-GitHub--ASJI5AsETsAFvfvodT_SwDXSAQ-BubGwzcgeuUgDTx0Kw4i3

**Comment by pkaminski** [comment]
Thanks for doing the extra research @abinoda, and for posting it here!  I find your argument persuasive and I'm now on board with adding some kind of automatic review re-requesting feature to Reviewable.  However, I'll need to think through the design carefully since Reviewable has its own notion of "user who needs to take action on the review" that doesn't align perfectly with GitHub's "requested reviewer".

Strawman design for feedback from Reviewable users:

Add a "Request reviewers: ..." toggle to the Publish options dropdown.  The toggle's state persists per-user, across all repos.  The request will include all users who will be considered _newly_ blocking by the review completion condition after the review is published, unless they are the PR author or they are already a requested reviewer.  You cannot add users manually to this list in the UI.  (Of course, you can always request reviewers via `+reviewer:@username` in any comment.)

I haven't tried it out but I think it should work reasonably well in a plain workflow.  It's not clear whether it will have surprising results if users are also manually manipulating the requested reviewers set.  It's also not clear whether having the setting be global-per-user is sufficient, or a more complex per-repo-per-user scheme (with per-repo admin-determined default?) is necessary.

**Comment by abinoda** [comment]
@pkaminski "automatic review re-requesting feature to Reviewable" — disclaimer again that I'm not familiar with Reviewable, but what do you mean by "automatic"? When I discussed this issue with GitHub, their idea (which I agree with) was to add a button that a PR Author could manually press. This would create a review request for one or more reviewers which is a better workflow than what PR Authors typically do, which is to write out a comment like "Hey @reviewer1 @reviewer2, I made changes, can you give this another reviewer?". Does what I'm saying make sense?

**Comment by pkaminski** [comment]
Reviewable doesn't subscribe to GitHub's UX philosophy of "hit this button, then that button, then another one if you remember" to accomplish a single logical task.  Instead, multiple actions are bundled into one button, with (usually persistent) options directing which extra actions to take.  The proposal here is to add such an option to the Publish button, which is Reviewable's master "I'm done with this round of the review" task button and the main way of (implicitly) changing the set of users blocking the review.  I think tacking the action of re-requesting users onto this task makes perfect sense, though it does need to be optional to allow for workflows that treat requested reviewers differently.

**Comment by abinoda** [comment]
@pkaminski Gotcha, makes sense. I'll have to try out Reviewable at some point to familiarize myself. In the meantime, feel free to reach out if you have any questions. I've been neck-deep in GitHub review and review request workflow and API stuff for the past 8 months so happy to provide any insights that may be helpful.

**Comment by gsuess** [comment]
@pkaminski 

> Reviewable has its own notion of "user who needs to take action on the review" that doesn't align perfectly with GitHub's "requested reviewer".

I am not quite sure what you are referring to there. So I am not sure what I am missing. Here is how I envision how this should work:

1. User A is the author of a Pull Request and requests a review from User B on GitHub (user B is now on the requested-reviews list of the PR on GitHub). 
1. User B adds a blocking discussion to the Pull request in Reviewable and rejects the PR on GitHub. The rejection implicitly removes the user from the list reviewers on GitHub. 
1. User A then fixes the issue and then goes to Reviewable and marks the discussion at Done. Upon publishing it on Reviewable, the PR is still blocking on the user B, because user B needs to acknowledge that the discussion is now resolved. 
1. Therefore when user A publishes the "Done" statement on Reviewable, Reviewable should add user B to the list of requested reviewers again.

There could potentially be situations where there are multiple reviewers, (i.e. user B and user C) who have blocking discussions on Reviewable. In this case Reviewable can push multiple reviewer requests to GitHub when the author publishes that the fact  that discussions are done.

**Comment by pkaminski** [comment]
I spent a few hours working on this and have something running, but it's not clear what's the best logic to use for picking which reviewers to request in non-trivial cases such as:

1. A reviewer responds to some but not all discussions where it's their turn.  Normally this would automatically make them unrequested in GitHub, but they would remain pending in Reviewable.  Should Reviewable force them to remain requested as well?

2. It's a reviewer's turn to answer comments, and they are requested in GitHub and pending in Reviewable.  For whatever reason, somebody manually cancels the review request in GitHub.  The PR author then posts more comments that require the reviewer's attention.  Should the reviewer be re-requested by Reviewable or left unrequested?

3. Two reviewers are participating in a review.  The author posts a new discussion, which makes both reviewers pending in Reviewable.  Should both be requested?  If yes, then suppose one reviewer answers the discussion, which makes _both_ non-pending.  That reviewer will automatically be unrequested, but should Reviewable unrequest the other too even though they did nothing?

4. A reviewer is requested manually, but isn't pending in Reviewable next time somebody publishes.  Should they be unrequested?  If not, can this situation be reliably distinguished from the previous one?

Basically, how tightly should Reviewable's pending participants and GitHub's requested reviewers be synchronized?  Lightly, so that GitHub's semantics are respected as much as possible and manual manipulation of requested reviewers is more likely to survive, or tightly, since GitHub's built-in semantics are inadequate and having separate-but-similar lists is confusing?

I'm not sure there's a good answer, in which case it's possible that I'm asking the wrong question.

**Comment by pkaminski** [comment]
Reviewable already lets you request reviewers (or change assignees) explicitly via directives in any comment (e.g., `+reviewer:@username`), so adding another button to do so wouldn't make much of a difference.  The real question (and what your suggestion doesn't address) is _who_ should be requested or assigned when the button is clicked (or toggle is on, or whatever).  If we can't figure out an automated way to determine this that would satisfy a significant majority of users then there's no point in having this feature.

I'm hoping to hear from some of the many people who 👍'd this issue what they'd like to see happen in the scenarios above.  What say you, folks?

**Comment by abinoda** [comment]
Gotcha, agree.d Would be great to hear from @gsuess and @veerenmandalia.

**Comment by gsuess** [comment]
Sorry, for late reply...

I agree with @pkaminski. I would like to suggest a similar approach as was done as with approving a PR when all files are reviewed and there are no blocking discussions are left. Here Reviewable, offers to approve the PR upon publishing the review.

> 1.  A reviewer responds to some but not all discussions where it's their turn. Normally this would automatically make them unrequested in GitHub, but they would remain pending in Reviewable. Should Reviewable force them to remain requested as well?

In this case I think that it's fine if Reviewable offers to unrequest it upon publishing the response, however its not an important feature here.

I would not consider it to be an invalid state when Reviewable claims that there are blocking discussions that the author has marked as resolved, but the reviewer is not requested in GitHub. In our case after all, the author bears the responsibility to ensure that the reviewer is explicitly requested on GitHub and we do not wish to entirely delegate that responsibility to Reviewable. Instead we would like to have reviewable make it easier for the author to manage these reviewer requests, while still having full control over which reviewers are requested.

> 2. It's a reviewer's turn to answer comments, and they are requested in GitHub and pending in Reviewable. For whatever reason, somebody manually cancels the review request in GitHub. The PR author then posts more comments that require the reviewer's attention. Should the reviewer be re-requested by Reviewable or left unrequested?

I would say, yes; Reviewable should offer to request another review from the reviewers. Again keeping in the spirit of helping the author to keep up with his or her responsibility to formally request reviews on Github.

> 3. Two reviewers are participating in a review. The author posts a new discussion, which makes both reviewers pending in Reviewable. Should both be requested? If yes, then suppose one reviewer answers the discussion, which makes both non-pending. That reviewer will automatically be unrequested, but should Reviewable unrequest the other too even though they did nothing?

Hmm, this is a tricky one. What do you mean by  "participating in a review"? Do you mean that this refers to the people who have been requested for review previously on that PR? If yes, then I would like to warn that this could become really messy fast: For example GitHub allows entire teams to be requested for review, but once anybody on a team has engaged in that review, the team is replaced by that individual. 

I do not think that it is necessary to care about who did reviews previously. What matters is only: Who's blocking discussions have been marked as "done" or otherwise marked as resolved on Reviewable. Now when these markings are published by the author, then Reviewable should offer to re-request the review from them.

> 4. A reviewer is requested manually, but isn't pending in Reviewable next time somebody publishes. Should they be unrequested? If not, can this situation be reliably distinguished from the previous one?

Yeah, as with point 3: I don't think Reviewable should take note on who was requested for review on GitHub previously. Instead it should just focus on who is needed for review in order to unblock the stuff that happens on Reviewable, regardless of what is going on beyond Reviewable.

**Comment by pkaminski** [comment]
So after thinking about this for a while, I believe the right thing to do is to sync the requested reviewers pretty tightly with Reviewable's pending reviewers because (i) the feature will be opt-in, so we don't need to worry about teams who find the binding too tight -- they just won't turn it on, and (ii) pending reviewers are already fully customizable, so the tighter the binding the better control that will also give you over requested reviewers.

With this in mind, here's the logic I'm proposing to apply when publishing a review with "sync requested reviewers" toggled on.  Define "old list" as the pending Reviewable reviewers immediately before the publication, and "new list" as the ones immediately after.  Then:
1.  Request a review from everyone on the new list -- including the publishing user, if present -- even if they were already on the old list.
2.  Unrequest a review from reviewers who were on the old list but aren't on the new one.  (This won't remove review requests for users who aren't on either list.)
3.  Override the above with any reviewer request directives in the comments being published.
4.  Apply all of the above _after_ the review is published, so it will overrule GitHub's "publishing satisfies a review request" if appropriate.

This logic would resolve the previously posited scenarios in the following way, assuming Reviewable's default pending reviewers algorithm is in effect:
1.  A partial review will result in the reviewer remaining requested.
2.  A manually canceled review requested will be reissued at the next publish (by anyone) if the reviewer is still pending.
3.  Both reviewers will be requested and both requests will be canceled.
4.  A manually requested reviewer will not be unrequested (unless they were also pending at publish time).

All of this can be tweaked to your heart's delight by customizing the `pendingReviewers` output from the custom review completion condition.  For example, if you'd like partial reviews to unrequest reviewers, you can remove all other users from any pending reviewers list that includes the PR author.  (Or you could be more subtle and only remove reviewers who have marked some files at the latest revision, or who are resolved on some discussions...  Your choice.)

Thoughts?

**Comment by pkaminski** [comment]
I went ahead and implemented the logic proposed above.  However, testing edge cases was nigh-impossible (there's only so many fake GitHub accounts I can create!) so please let me know if you run into weirdness before I announce the feature more widely.  It's now live, with docs [here](https://docs.reviewable.io/#/reviews?id=sync-requested-reviewers) (including a shout out to Pull Reminders 😄).

**Comment by gsuess** [comment]
I can't really see any change with regards to the issue have described really. Is it already released?

I had a open discussion from a code reviewer. Pushed the changes he requested. I have marked the discussion as "Done" in reviewable and published my remark. The review was not re-requested.

**Comment by pkaminski** [comment]
The feature was live two weeks ago.  It's opt-in, so look for a checkbox in the Publish options dropdown.
